src/main/scala/org/broadinstitute/dsde/firecloud/model/ModelJsonProtocol.scala

```diff
@@ -131,21 +140,71 @@ object ModelJsonProtocol {
         case None => None
         case _ => throw DeserializationException("unexpected json type for " + SIZE)
       }
-      LibrarySearchParams(term, from, size)
+      val fields: Map[String, Seq[String]] = data.getOrElse(SEARCH_FIELDS, None) match {
+        case some:JsObject => some.convertTo[Map[String,Seq[String]]]
+        case None => Map.empty
+      }
+      val aggs = data.getOrElse(FIELD_AGGREGATIONS, None) match {
+        case some: JsArray => some.convertTo[Seq[String]]
+        case None => Seq.empty
+        case _ => throw DeserializationException("unexpected json type")
+      }
+      val maxAggs = data.getOrElse(MAX_AGGREGATIONS, None) match {
+        case JsNumber(x) => Some(x.intValue)
+        case None => None
+        case _ => throw DeserializationException("unexpected json type")
+      }
+      LibrarySearchParams(term, fields, aggs, maxAggs, from, size)
+    }
+  }
+
+  implicit object impESLogicType extends JsonFormat[ESLogicType] {
+    override def write(inputmap: ESLogicType): JsValue = inputmap match {
+      case must : ESMust => must.toJson
+      case should : ESShould => should.toJson
+      case _ => throw new SerializationException("unexpected QueryMap type")
+    }
+
+    override def read(json: JsValue): ESLogicType = {
+      json.asJsObject.fields.keys.head match {
+        case "must" => impESMust.read(json)
+        case "should" => impESShould.read(json)
+        case _ => throw DeserializationException("unexpected json type")
+      }
     }
   }
 
   implicit object impQueryMap extends JsonFormat[QueryMap] {
     override def write(inputmap: QueryMap): JsValue = inputmap match {
       case matchall : ESMatchAll => matchall.toJson
-      case wildcard : ESWildcard => wildcard.toJson
+      case amatch : ESMatch => amatch.toJson
+      case term : ESTerm => term.toJson
+      case bool : ESBool => bool.toJson
       case _ => throw new SerializationException("unexpected QueryMap type")
     }
 
     override def read(json: JsValue): QueryMap = {
       json.asJsObject.fields.keys.head match {
         case "match_all" => impESMatchAll.read(json)
-        case "wildcard" => impESWildcard.read(json)
+        case "match" => impESMatch.read(json)
+        case "term" => impESTerm.read(json)
+        case "bool" => impESBool.read(json)
+        case _ => throw DeserializationException("unexpected json type")
+      }
+    }
+  }
+
+  implicit object impESPropertyFields extends JsonFormat[ESPropertyFields] {
+    override def write(input: ESPropertyFields): JsValue = input match {
+      case estype : ESType => estype.toJson
+      case esaggtype : ESAggregatableType => esaggtype.toJson
+      case _ => throw new SerializationException("unexpected ESProperty type")
+    }
+
+    override def read(json: JsValue): ESPropertyFields = {
+      json.asJsObject.fields.keys.size match {
+        case 1 => ESTypeFormat.read(json)
+        case 2 => ESAggregatableTypeFormat.read(json)
```

Comment:
do you prefer this?
    override def read(json: JsValue): ESPropertyFields = {
      if (json.asJsObject.fields.contains("fields")) {
        ESAggregatableTypeFormat.read(json)
      } else {
        ESTypeFormat.read(json)
      }
    }